lib/Dialect/FIRRTL/Transforms/Vectorization.cpp

```diff
@@ -69,7 +69,70 @@ class VectorCreateToLogicElementwise : public mlir::RewritePattern {
     return failure();
   }
 };
-} // namespace
+
+template <typename OpTy, typename ResultOpType>
+class ChainedReducer : public mlir::RewritePattern {
+public:
+  ChainedReducer(MLIRContext *context)
+      : RewritePattern(OpTy::getOperationName(), 0, context) {}
+
+  bool recurse(OpTy root, Value vec, DenseSet<size_t> &indexes) const {
```

Comment:
Could this be iterative instead? Looks like it might recurse based on number of elements in the vector?